### PR TITLE
ref(metrics-indexer): Turn process_messages into MessageProcessor

### PR DESCRIPTION
**Pull request by MeredithAnya**
**context:**
addressing the comments in https://github.com/getsentry/sentry/pull/37745 specifically this comment https://github.com/getsentry/sentry/pull/37745#discussion_r949776817.

This change will allow me to initialize the indexer _not_ on every batch, but instead on when the strategy is created. If this would be recreating the indexer instance too often, we could move it farther up (aka when the factory is created vs the strategy). 

My idea is that in https://github.com/getsentry/sentry/pull/37745 we can add a `indexer_db_options` to the config which can be used to initialize the indexer. 

Something like:

```python
STORAGE_TO_INDEXER: Mapping[IndexerStorage, StringIndexer] = {
    IndexerStorage.CLOUDSPANNER: CloudSpannerIndexer,
    IndexerStorage.POSTGRES: PostgresIndexer,
    IndexerStorage.MOCK: MockIndexer,
}

class MessageProcessor:
    def __init__(self, config: MetricsIngestConfiguration):
        self._config = config
        self._indexer = STORAGE_TO_INDEXER[config.db_backend](**config.indexer_db_options)

### CODE REVIEW

File: src/sentry/sentry_metrics/consumers/indexer/parallel.py [hunk 1]
```diff
@@ -112,8 +111,9 @@ def create_with_partitions(
         commit: Callable[[Mapping[Partition, Position]], None],
         partitions: Mapping[Partition, int],
     ) -> ProcessingStrategy[KafkaPayload]:
+        message_processor = MessageProcessor(self.__config.use_case_id)
         parallel_strategy = ParallelTransformStep(
-            partial(process_messages, self.__config.use_case_id),
+            message_processor.process_messages,
```
Comment (fpacifici) [inline src/sentry/sentry_metrics/consumers/indexer/parallel.py:116]: nit: Please just initialize the processor here. You do not need the `message_processor` variable.

**Review by fpacifici — APPROVED** [review]
Thanks

File: tests/sentry/sentry_metrics/test_multiprocess_steps.py [hunk 2]
```diff
@@ -275,7 +277,7 @@ def test_process_messages() -> None:
     last = message_batch[-1]
     outer_message = Message(last.partition, last.offset, message_batch, last.timestamp)
 
-    new_batch = process_messages(use_case_id=UseCaseKey.RELEASE_HEALTH, outer_message=outer_message)
+    new_batch = process_messages(outer_message=outer_message)
```
Comment (fpacifici) [inline tests/sentry/sentry_metrics/test_multiprocess_steps.py:280]: Please instantiate the message processor in the test and use that instance. Now calling the process_messages method directly is not relevant anymore.

**Review by untitaker — APPROVED** [review]
besides filippo's comments i don't think there's anything I'd change

**Pull request merged**